content/src/tnips/tnip_2.md

```diff
@@ -0,0 +1,279 @@
+
+* *e* : The epoch boundary.
+* *e<sub>t</sub>* : "Epoch boundary time" is the UNIX timestamp used to quantify the epoch boundary (currently expected to be 24-hours after the epoch starts).
+* *e<sub>r</sub>* : "Epoch boundary round" is the last round of consensus committed by all [CVVs](#committee-voting-validators-cvvs) in the current epoch.
+* *e<sub>n</sub>* : "Epoch boundary number" is the number of epochs, starting at 0 and incrementing +1 with every new epoch.
+* *e<sub>m</sub>* : "Epoch boundary max" is the maximum amount of time (in seconds) that the pending committee should wait before falling back to a previous execution result to include in their genesis certificate.
+
+NOTE: The timestamp for *e<sub>r</sub>* must be within [*e<sub>t</sub>* , *e<sub>t</sub>* + *e<sub>m</sub>*).
+
+##### Signed Execution Result
+
+The sealed header of the last block produced while executing *e<sub>r</sub>*. The header's hash is signed by the node’s BLS12-381 private key. The signed execution result includes the sealed header and the signed hash of the sealed header. Peers verify signed execution results by taking the SHA-256 hash of the header and comparing it to the signature.
+
+#### Recovering from failed execution at *e*
+
+Byzantine nodes must not prevent the network from closing an epoch or a new committee from taking over consensus. If the (pending) current committee C fails to receive a quorum of signed execution results at *e* to start the epoch, then the certified block in the last DAG commit is used to start the epoch. The new committee C must use this leader certificate’s aggregate signatures to generate C<sub>n+1</sub> since the closing committee failed to propose this committee. Nodes must be penalized in the first round of the new epoch by slashing stake if they failed to attest the epoch boundary. The penalty must be applied in addition to the node’s other opening epoch responsibilities. The amount of stake to deduct from the validator’s staked balance will be determined through social governance.
+
+#### Closing Epoch Responsibilities
+
+[CVVs](#committee-voting-validators-cvvs) must perform closing epoch responsibilities to facilitate a successful transfer of consensus voting privileges to the next committee.
+
+##### Executing the last output
+
+[CVVs](#committee-voting-validators-cvvs) reach consensus and commit rounds to a local DAG until a certain UNIX timestamp is reached (et). Once e<sub>t</sub> is reached, [CVVs](#committee-voting-validators-cvvs) have up to 60 seconds (or enough time to ensure execution attempt, even if extra rounds needed for commit) to commit a leader and execute the output from consensus. While executing the last output, [CVVs](#committee-voting-validators-cvvs) must update the consensus registry contract with committee’s updates using a system call to `concludeEpoch` on the consensus registry contract at 0x0111  The required committee updates are:
```

Comment:
This is really helpful for me to read as a spec for the planned future PR implementing these system calls. If I'm reading this right, this is the concept we discussed a couple weeks back where all validators must insert a final `system_call::concludeEpoch()` transaction as part of their execution. I remember you mentioning this happens in the engine so I will be sure to keep this in mind for my future work. Thanks!